### PR TITLE
[mta] implement `_foreach_pow`

### PR DESCRIPTION
**Pull request by crcrpar**
Mainly for foreach path of `Adam` and `AdamW`

rel: https://github.com/pytorch/pytorch/issues/58833

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 1]
```diff
@@ -179,9 +180,9 @@ __device__ __forceinline__ void pointwise_op_scalar(
 //
 // Binary Functors
 //
-template<typename T, int depth, int r_args_depth, int res_arg_index>
+template<typename T, int depth, int r_args_depth, int res_arg_index, bool disable_opmath_type=false>
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: What is this change for?

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 2]
```diff
@@ -547,5 +548,19 @@ struct TernaryOpScalarFunctor {
...
 
+template <typename T>
+struct power_functor {
+  C10_DEVICE T operator()(const T& a, const T& b) const {
+    return pow(a, b);
+  }
+};
+
+template <typename T>
+struct alt_power_functor {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:559]: Why do you need this one?

File: aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu [hunk 3]
```diff
@@ -103,12 +120,13 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
...
     }                                                                                                                    \
                                                                                                                          \
-    return FUNCTION<OP>(tensors, scalars);                                                                               \
+    return FUNCTION<OP, DISABLE_OPMATH_TYPE>(tensors, scalars);                                                          \
 }
 
-FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, add, std::plus, /*div_op*/ false);
-FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, mul, std::multiplies, /*div_op*/ false);
-FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, div, std::divides, /*div_op*/ true);
+FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, add, std::plus, /*div_op*/ false, false);
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:126]: Comments for what the boolean arg is for readability.

**Review by albanD — COMMENTED** [review]
Sounds good. Mainly readability questions in the kernel code.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 2 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:559]: `multi_tensor_apply` assumes that `TensorList` is SelfArgument but `_foreach_pow(scalar, tensor_list)` doesn't satisfy the assumption. so I needed to exchange the order of arguments before passing them to `pow` function

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 1 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: I saw numerical differences of the outputs from `multi_tensor_apply` and the native for fp16 and bf16 inputs (`opmath_t` is float32 for float16 and bfloat16, scalar_type for the other scalar_type's) so I concluded power kernels don't upcast low-precision numbers.
Is my understanding correct?

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 2 above)
Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:559]: Can you make the name more explicit then? `reverse_power_functor` would match better the usual `rmethod` that exist in python world.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: I'm not sure to understand? You don't want these to run with high precision intermediary? Why?

Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: To implement the same behavior as the native kernels.
e.g. https://github.com/pytorch/pytorch/blob/5b2afaaca8ca36c3029e22664dd63ee1cf2461b4/aten/src/ATen/native/cuda/PowKernel.cu#L124-L141 doesn't seem to use float32 for float16 and bfloat16 inputs.

Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: cc @ngimel is this a bug or a feature? :)

Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: From offline chat:
- There shouldn't be any measurable numerical difference between the two for a single op.
- Is there any measurable perf difference between upcasting or not? If so, let's go with the fastests.
- If not, we should just go with the simplest implementation.

Comment (albanD) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:183]: from offline discussion, pow_scalar_tensor_impl call into pow_ which does convert to float before calling std::pow.
So we should be doing the same here in the end.

File: aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu [hunk 4]
```diff
@@ -109,6 +124,7 @@ std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList te
 FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, add, std::plus, /*div_op*/ false);
 FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, mul, std::multiplies, /*div_op*/ false);
 FOREACH_BINARY_OP_SCALARLIST(all_types_complex_bool_half_bfloat16, div, std::divides, /*div_op*/ true);
+FOREACH_BINARY_OP_SCALARLIST(all_types_complex_half_bfloat16, pow, power_functor, /*div_op*/ true);
```
Comment (albanD) [inline aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:127]: Why is this div_op true?

**Review by albanD — COMMENTED** [review]
Thanks for the cleanup! Just a small question.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 5]
```diff
@@ -547,5 +548,19 @@ struct TernaryOpScalarFunctor {
   }
 };
 
+template <typename T>
+struct power_functor {
+  C10_DEVICE T operator()(const T& a, const T& b) const {
+    return pow(a, b);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:554]: what's the name resolution here?

Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh]: I meant cuda's pow. updated it to `at::native::pow_`

**Review by albanD — APPROVED** [review]
Thanks for the comment!
Sounds good!

**Pull request closed without merging**